frontend/src/metabase/query_builder/actions.js

```diff
@@ -20,11 +21,81 @@ import { getParameters } from "./selectors";
 const Metabase = new AngularResourceProxy("Metabase", ["db_list_with_tables", "db_fields", "dataset", "table_query_metadata"]);
 const User = new AngularResourceProxy("User", ["update_qbnewb"]);
 
+import { parse as urlParse } from "url";
+
+export const SET_CURRENT_STATE = "metabase/qb/SET_CURRENT_STATE";
+const setCurrentState = createAction(SET_CURRENT_STATE);
+
+export const POP_STATE = "metabase/qb/POP_STATE";
+export const popState = createThunkAction(POP_STATE, (location) =>
+    async (dispatch, getState) => {
+        const { card } = getState().qb;
+        if (location.state && location.state.card) {
+            if (!angular.equals(card, location.state.card)) {
+                console.log("QB POP", location);
+                dispatch(setCardAndRun(location.state.card, false));
+                dispatch(setCurrentState(location.state));
+            } else {
+                console.log("QB SKIP POP")
+            }
+        }
+    }
+);
+
+function updateUrl(dispatch, getState, card, isDirty=false, replaceState=false) {
```

Comment:
Hmm... couldn't we make this an actual action? So we can dispatch it and have access to dispatch and getState without having to pass it in? The other args can just be part of the payload right?